lib/client/db/postgres/repl/repl.go

```diff
@@ -0,0 +1,288 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package repl
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"strings"
+
+	"github.com/gravitational/trace"
+	"github.com/jackc/pgconn"
+	"golang.org/x/term"
+
+	"github.com/gravitational/teleport"
+	clientproto "github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/lib/asciitable"
+	"github.com/gravitational/teleport/lib/defaults"
+)
+
+type REPL struct {
+	connConfig *pgconn.Config
+	client     io.ReadWriteCloser
+	serverConn net.Conn
+	route      clientproto.RouteToDatabase
+	term       *term.Terminal
+	commands   map[string]*command
+}
+
+func New(ctx context.Context, client io.ReadWriteCloser, serverConn net.Conn, route clientproto.RouteToDatabase) (*REPL, error) {
+	config, err := pgconn.ParseConfig(fmt.Sprintf("postgres://%s", hostnamePlaceholder))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	config.User = route.Username
+	config.Database = route.Database
+	config.ConnectTimeout = defaults.DatabaseConnectTimeout
+	config.RuntimeParams = map[string]string{
+		applicationNameParamName: applicationNameParamValue,
+	}
+	config.TLSConfig = nil
+
+	// Provide a lookup function to avoid having the hostname placeholder to
+	// resolve into something else. Note that the returned value won't be used.
+	config.LookupFunc = func(_ context.Context, _ string) ([]string, error) {
+		return []string{hostnamePlaceholder}, nil
+	}
+	config.DialFunc = func(_ context.Context, _, _ string) (net.Conn, error) {
+		return serverConn, nil
+	}
+
+	return &REPL{
+		connConfig: config,
+		client:     client,
+		serverConn: serverConn,
+		route:      route,
+		term:       term.NewTerminal(client, ""),
+		commands:   initCommands(),
+	}, nil
+}
+
+// Run starts and run the PostgreSQL REPL session. The provided context is used
+// to interrupt the execution and clean up resources.
+func (r *REPL) Run(ctx context.Context) error {
+	pgConn, err := pgconn.ConnectConfig(ctx, r.connConfig)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer pgConn.Close(context.TODO())
```

Comment:
Should we use the parent context or an independent, time-bound context instead? The docs for `pgConn.Close()` state:

```
Close closes a connection. It is safe to call Close on a already closed connection. Close attempts a clean close by sending the exit message to PostgreSQL. However, this could block so ctx is available to limit the time to wait. The underlying net.Conn.Close() will always be called regardless of any other errors.
```

Reply:
ngl I needed to figure out which context to use here. @greedy52 is correct. If we use `context.Background()` directly, it won't have the effect we want (which is to send the terminate command).

What do you think about this? Should we create another context here (with a timeout) or receive it from the caller (like a teardown context)?

Reply:
My main worry is having this hanging indefinitely, blocking further cleanups. I feel like giving this a few seconds (3? 5?) for a cleanup is enough and wouldn't impact the teardown too much. I'm not sure if using teardown context will lead to clear shutdown, but that is an option as well.